File: src/core/audio/content_addressing.py
"""T-W4-022 音频素材内容寻址（架构 v2 §4.6 / 宪法 D3）.

D3 内容寻址身份：同一内容必得同一 id。音频素材的「内容」由四个维度唯一决定：
    audio_id = H(text, voice_profile, speed, engine_digest)

为什么纳入 engine_digest：不同 TTS 引擎合成质量不同，即使文本/音色/语速相同，
音频内容也不同——engine 是内容寻址的一部分（与 T-W4-011 compute_content_id
同语义，确保产线 id 与总线 id 一致，便于回溯）。

为什么 speed 而非 wpm：speed 是学段配置的语速（wpm），此处用通用名 speed
表达「语速参数」，落值为 wpm 整数。命名解耦避免音频层硬绑 TTS 术语。

本模块与 T-W4-011 的 compute_content_id 保持同公式（text|voice|wpm|engine），
使 produce_audio 产出的 audio_id 与 tts_synthesize 返回的 content_id 一致——
码内可回溯（架构 §4.6）。
"""
from __future__ import annotations

import hashlib


def compute_audio_content_id(
    text: str, voice_profile: str, speed: int, engine_digest: str
) -> str:
    """音频素材内容寻址 id（D3）.

    公式：sha256("{text}|{voice_profile}|{speed}|{engine_digest}") 前 32 位 hex。

    Args:
        text: 已剥离 PII 的合成文本（D7：PII 剥离由上层调用方负责）。
        voice_profile: 音色名（如 female_standard，对应 voice_profiles.yaml）。
        speed: 语速 wpm（学段配置：L=120/M=140/H=160）。
        engine_digest: 引擎标识（如 'mock'；生产为真实适配器注册名）。

    Returns:
        32 位 hex 字符串（与 T-W4-011 compute_content_id 同公式同结果）。

    Notes:
        相同输入返回相同 id；任何参数变更（text/voice/speed/engine）产生新 id。
    """
    payload = f"{text}|{voice_profile}|{speed}|{engine_digest}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:32]


def compute_content_hash(audio: bytes) -> str:
    """音频字节流的规范化哈希（对象存储去重与完整性校验用）.

    为什么与 audio_id 分离：audio_id 是「输入参数寻址」（同参数同 id，命中缓存）；
    content_hash 是「产物字节寻址」（同字节同哈希，用于存储层去重与校验完整性）。
    两者维度不同——同 audio_id 必同 content_hash（确定性合成），但同 content_hash
    可能来自不同参数（理论上不会，但分离语义更清晰）。

    Returns:
        'sha256:' + 64 位 hex（与 gate validators 的 _canonical_hash 前缀风格一致）。
    """
    return "sha256:" + hashlib.sha256(audio).hexdigest()


__all__ = ["compute_audio_content_id", "compute_content_hash"]

File: core/audio/content_addressing.go
// Package audio 承载音频资产域（T-W4-022/024/026 的 Go 重锚定，冻结实现
// src/core/audio/，架构 v2 §4.6 / S5）——TTS（core/ai/tts）下游的音频资产
// 管理面：内容寻址、点读时间轴、音频产线、卷面贴码、听力端到端数据面、
// 限次播放。
//
// 职责边界：
//   - 音频内容寻址（D3）：冻结公式 sha256(text|voice|speed|engine) 前 32 hex
//     的跨语言地面真值锚点（content_addressing.go）；Go TTS 总线产物 id 的
//     修复面公式（完整摘要、键含剥离后文本）在 core/ai/tts 单点实现，产线
//     对总线 id 透传（总线 id == 产线 id == 存储寻址，三位一体由单一公式
//     实现保证）；
//   - 点读/播放/贴码/听力均为纯数据面或端口注入面：对象存储写入、播放计数
//     存储、音频过门验证器全部以接口注入（与 TTSEngine 同模式的副作用边界），
//     测试以确定性假件 hermetic 覆盖，生产替换 MinIO/Redis/验证器装配适配器；
//   - QR SVG 位图生成复用 core/render.GenerateQRSVG——零新依赖约束下该实现
//     是显式骨架（ErrQRSVGNotImplemented），本包不静默降级（见 qr.go）；
//   - 宪法 A5/X6：不 import 学科包/学段包；学段参数经 core/ai/tts 配置表注入；
//     宪法 D7：PII 剥离在 TTS 总线内 fail-closed，本包不感知 PII 语义。
//
// 冻结对照：content_addressing.py / point_read.py / producer.py /
// qr_generator.py / listening_e2e.py / player_service.py。
package audio

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// audioIDHexLen 音频内容寻址 id 截取长度（冻结实现 hashlib.sha256(...).
// hexdigest()[:32]，32 位 hex）.
const audioIDHexLen = 32

// contentHashPrefix 内容哈希前缀（与 gate validators 的 _canonical_hash
// 前缀风格一致）.
const contentHashPrefix = "sha256:"

// ComputeAudioContentID 计算音频素材内容寻址 id（D3；冻结实现
// compute_audio_content_id 对齐）。
//
// 公式：sha256("{text}|{voice_profile}|{speed}|{engine_digest}") 前 32 位 hex。
//
//   - text：已剥离 PII 的合成文本（D7：PII 剥离由上层调用方负责）；
//   - voiceProfile：音色名（如 female_standard，对应 voice_profiles 表）；
//   - speed：语速 wpm（学段配置：L=120/M=140/H=160）；以通用名 speed 表达
//     「语速参数」、落值为 wpm 整数——命名解耦避免音频层硬绑 TTS 术语；
//   - engineDigest：引擎标识（如 "mock"；生产为真实适配器注册名）。
//
// 相同输入返回相同 id；任何参数变更（text/voice/speed/engine）产生新 id。
// 与冻结实现逐字节互验（golden 测试锁定采样值）。
//
// 与 Go TTS 总线 id 的关系（显式偏离，如实声明）：core/ai/tts 的产物 id 采用
// 修复后的公式（剥离后文本|学段|音色|voice_id|语速|引擎，完整 64 hex——冻结
// 实现截断 32 hex 的碰撞风险已在 T-W5-015 修复）。本函数承担：①冻结 id 体系
// （历史数据/审计回溯）在 Go 侧的可复现入口；②「音频 id 与文本 id 同公式族
// （text|voice|wpm|engine）」的语义证明。产线（producer.go）对 Go 总线 id
// 透传，不再持第二套公式做运行时防御断言——冻结实现的「双实现互验」在 Go 侧
// 收敛为本文件的黄金测试互验。
func ComputeAudioContentID(text, voiceProfile string, speed int, engineDigest string) string {
	payload := fmt.Sprintf("%s|%s|%d|%s", text, voiceProfile, speed, engineDigest)
	sum := sha256.Sum256([]byte(payload))
	return hex.EncodeToString(sum[:])[:audioIDHexLen]
}

// ComputeContentHash 计算音频字节流的规范化哈希（冻结实现
// compute_content_hash 对齐）：'sha256:' + 64 位 hex。
//
// 为什么与 audio_id 分离：audio_id 是「输入参数寻址」（同参数同 id，命中
// 缓存）；content_hash 是「产物字节寻址」（同字节同哈希，用于存储层去重与
// 完整性校验）。两者维度不同——同 audio_id 必同 content_hash（确定性合成）。
func ComputeContentHash(audio []byte) string {
	sum := sha256.Sum256(audio)
	return contentHashPrefix + hex.EncodeToString(sum[:])
}
